drivers/src/dma.rs

```diff
@@ -164,13 +181,23 @@ impl Dma {
             return Err(CaliptraError::DRIVER_DMA_FIFO_UNDERRUN);
         }
 
-        read_data.chunks_mut(4).for_each(|word| {
-            let ptr = dma.read_data().ptr as *mut u8;
-            // Reg only exports u32 writes but we need finer grained access
-            unsafe {
-                ptr.copy_to_nonoverlapping(word.as_mut_ptr(), word.len());
+        let read_data_ptr = dma.read_data().ptr as *const u8;
+
+        // Process all complete 4-byte chunks
+        for chunk in read_data.chunks_exact_mut(4) {
+            let value = unsafe { read_volatile(read_data_ptr as *const u32) };
+            chunk.copy_from_slice(&value.to_le_bytes());
+        }
+
+        // Process remaining bytes (1, 2, or 3)
+        let remainder = read_data.len() % 4;
+        if remainder > 0 {
+            let start = read_data.len() - remainder;
+            let last_chunk = &mut read_data[start..];
+            for byte in last_chunk.iter_mut() {
+                *byte = unsafe { read_volatile(read_data_ptr) };
```

Comment:
Is this how the hardware works?

Since it is a 32-bit register underneath mapping to a 32-bit LSU, I would have expected we would do one last 32-bit read and then discard leading or trailing zeros to get the remainder bytes.

Similar for writing.

Reply:
I don't know what the hardware does. The emulator does panic on fifo sizes that aren't a multiple of 4 so the driver might as well bail out as I don't think we need this.